src/main/java/org/gaul/s3proxy/S3ProxyHandler.java

```diff
@@ -384,12 +374,57 @@ public final void doHandle(HttpServletRequest baseRequest,
 
             try {
                 authHeader = new S3AuthorizationHeader(headerAuthorization);
+                //whether v2 or v4 (normal header and query)
             } catch (IllegalArgumentException iae) {
                 throw new S3Exception(S3ErrorCode.INVALID_ARGUMENT, iae);
             }
             requestIdentity = authHeader.identity;
         }
 
+        if (hasXAmzDateHeader) { //format diff between v2 and v4
+            String dateString = request.getHeader("x-amz-date");
+            switch (authenticationType) { //proxys3 auth type support
+            case AWS_V2:
+                long date = parseRfc2616(dateString);
+                isTimeskewed(date);
+                break;
+            case AWS_V4:
+                logger.debug("into process v4");
+                date = parseIso8601(dateString);
+                isTimeskewed(date);
+                break;
+            case AWS_V2_OR_V4:
+                logger.debug("into process v2v4");
+                if (authHeader.authenticationType ==
+                    AuthenticationType.AWS_V2) {
+                    date = parseRfc2616(dateString); //todo all rfc2616 variance
+                    isTimeskewed(date);
+                } else if (authHeader.authenticationType ==
+                    AuthenticationType.AWS_V4) {
+                    date = parseIso8601(dateString);
+                    isTimeskewed(date);
```

Comment:
Instead of calling `isTimeSkewed` several times could you do this once at the end?